### DISCUSSION TITLE
Using mypy to typecheck instantiation of classes mapped with decorators fails with 2.0 syntax

### DISCUSSION THREAD

**Discussion by mhwaage** [sqlalchemy/sqlalchemy#8686] [Usage Questions]
### Describe the bug

Classes that are mapped using a registry + decorator, rather than the DeclarativeBase, are not correctly type checked by Mypy.

### To Reproduce

```python
from sqlalchemy.orm import registry, Mapped, mapped_column, DeclarativeBase

reg = registry()

@reg.mapped  # Also reproduces with @reg.mapped_as_dataclass
class Foo:
    __tablename__ = "Foo"
    val: Mapped[int] = mapped_column(primary_key=True)

x = Foo(val=7)  # error: Unexpected keyword argument "val" for "Foo"

class Base(DeclarativeBase):
    pass

class Bar(Base):
    __tablename__ = "Bar"
    val: Mapped[int] = mapped_column(primary_key=True)

y = Bar(val=7)  # OK
```

### Error

When running mypy on the reproducing code, mypy finds one error:
```
error: Unexpected keyword argument "val" for "Foo"
``` 

### Versions

- OS: Win10
- Python: 3.10
- SQLAlchemy: 2.0.0b1 (also checked on github main@05d5f90)
- mypy: 0.982
- Database: N/A
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): N/A

**Comment by CaselIT** [comment]
Hi,

The case of `@reg.mapped` I don't think it's solvable without using the mypy plugin.
The case of `@reg.mapped_as_dataclass` is not supperted yet by mypy as documented, since it does not support https://peps.python.org/pep-0681/. pyright / pylance will work with it (and I think it will also work with the mypy plugin)

There is not much that sqlalchemy can do here

**Comment by mhwaage** [comment]
Thank you for the reply! I guess the best way forward in my case is to shift to a declarative base.

**Comment by CaselIT** [comment]
Let's move to a discussion since it may be useful to others
